File: apps/web/lib/api/access-reviews.ts
import { z } from "zod";
import { apiRequest } from "./client";

export const reviewScopeSchema = z.object({
  target_type: z.enum(["matter", "ip_docket"]), target_id: z.string(), target_title: z.string(), access_policy_version: z.number().int(),
  grants: z.array(z.object({ id: z.string(), record_version: z.number().int(), subject_type: z.enum(["membership", "team"]), subject_id: z.string(), subject_label: z.string(), effective_from: z.string().nullable(), expires_at: z.string().nullable(), reason: z.string().nullable() })),
});
export const campaignSchema = z.object({
  id: z.string(), title: z.string(), reason: z.string(), trigger: z.string(), status: z.enum(["open", "finalized"]), version: z.number().int(), creator_user_id: z.string(), created_at: z.string(), finalized_at: z.string().nullable(), snapshot: reviewScopeSchema,
  decisions: z.array(z.object({ grant_id: z.string(), decision: z.enum(["keep", "revoke"]), reason: z.string(), reviewer_user_id: z.string(), reviewer_membership_id: z.string(), created_at: z.string() })),
});
export type Campaign = z.infer<typeof campaignSchema>;
export type ReviewScope = z.infer<typeof reviewScopeSchema>;
export type ReviewTarget = ReviewScope["target_type"];
export type ReviewTrigger = "periodic" | "client_team_change" | "ethical_wall_change" | "portal_inactivity" | "employee_change" | "counsel_completion" | "incident";
const base = "/api/access-reviews";
const path = (id: string) => `${base}/${encodeURIComponent(id)}`;

export async function listReviewTargets(kind: ReviewTarget, q: string, afterId: string | null, signal?: AbortSignal) {
  const query = new URLSearchParams({ kind, q });
  if (afterId) query.set("after_id", afterId);
  return z.object({ targets: z.array(z.object({ id: z.string(), title: z.string() })), next_after_id: z.string().nullable() }).parse(await apiRequest(`${base}/targets?${query}`, { signal }));
}
export async function fetchReviewScope(kind: ReviewTarget, targetId: string, signal?: AbortSignal) {
  return reviewScopeSchema.parse(await apiRequest(`${base}/scope?${new URLSearchParams({ kind, target_id: targetId })}`, { signal }));
}
export async function listAccessReviews(beforeId: string | null, signal?: AbortSignal) {
  const query = new URLSearchParams();
  if (beforeId) query.set("before_id", beforeId);
  return z.object({ campaigns: z.array(campaignSchema), next_before_id: z.string().nullable() }).parse(await apiRequest(`${base}?${query}`, { signal }));
}
export async function fetchAccessReview(id: string, signal?: AbortSignal) {
  return campaignSchema.parse(await apiRequest(path(id), { signal }));
}
export async function createAccessReview(scope: ReviewScope, title: string, reason: string, trigger: ReviewTrigger) {
  return campaignSchema.parse(await apiRequest(base, { method: "POST", body: { target_type: scope.target_type, target_id: scope.target_id, expected_access_policy_version: scope.access_policy_version, title, reason, trigger } }));
}
export async function decideAccessReview(campaign: Campaign, grantId: string, decision: "keep" | "revoke", reason: string) {
  return campaignSchema.parse(await apiRequest(`${path(campaign.id)}/decisions`, { method: "POST", body: { expected_version: campaign.version, grant_id: grantId, decision, reason } }));
}
export async function finalizeAccessReview(campaign: Campaign) {
  return campaignSchema.parse(await apiRequest(`${path(campaign.id)}/finalize`, { method: "POST", body: { expected_version: campaign.version } }));
}

File: apps/api/src/caseops_api/schemas/access_reviews.py
"""Bounded internal record-access review contracts; no new grant authority."""

from datetime import UTC, datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

TargetType = Literal["matter", "ip_docket"]


class Command(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)


class CampaignCreate(Command):
    target_type: TargetType
    target_id: str = Field(min_length=1, max_length=36)
    title: str = Field(min_length=3, max_length=200)
    reason: str = Field(min_length=5, max_length=1000)
    trigger: Literal[
        "periodic",
        "client_team_change",
        "ethical_wall_change",
        "portal_inactivity",
        "employee_change",
        "counsel_completion",
        "incident",
    ]
    expected_access_policy_version: int = Field(ge=0)


class VersionCommand(Command):
    expected_version: int = Field(ge=1)


class ReviewDecision(VersionCommand):
    grant_id: str = Field(min_length=1, max_length=36)
    decision: Literal["keep", "revoke"]
    reason: str = Field(min_length=5, max_length=1000)


class DatedRecord(BaseModel):
    @field_validator(
        "created_at",
        "finalized_at",
        "effective_from",
        "expires_at",
        mode="after",
        check_fields=False,
    )
    @classmethod
    def utc_dates(cls, value):
        return value.replace(tzinfo=UTC) if value is not None and value.tzinfo is None else value


class GrantSnapshot(DatedRecord):
    id: str
    record_version: int
    subject_type: Literal["membership", "team"]
    subject_id: str
    subject_label: str
    effective_from: datetime | None
    expires_at: datetime | None
    reason: str | None


class ScopeSnapshot(BaseModel):
    target_type: TargetType
    target_id: str
    target_title: str
    access_policy_version: int
    grants: list[GrantSnapshot]


class DecisionRecord(DatedRecord):
    grant_id: str
    decision: Literal["keep", "revoke"]
    reason: str
    reviewer_user_id: str
    reviewer_membership_id: str
    created_at: datetime


class CampaignRecord(DatedRecord):
    id: str
    title: str
    reason: str
    trigger: str
    status: Literal["open", "finalized"]
    version: int
    creator_user_id: str
    created_at: datetime
    finalized_at: datetime | None
    snapshot: ScopeSnapshot
    decisions: list[DecisionRecord]


class CampaignPage(BaseModel):
    campaigns: list[CampaignRecord]
    next_before_id: str | None


class TargetOption(BaseModel):
    id: str
    title: str


class TargetPage(BaseModel):
    targets: list[TargetOption]
    next_after_id: str | None
